### PR TITLE
refactor: Improve project structure and add testing

### LINKED ISSUE #2: [Feature Request] Scrape course data
### Overview

Create a web scraper to extract course data from the UofA course planner website: https://courseplanner.uni.adelaide.edu.au. 

### Details

- Scrape UofA course data using their API.
- Schedule regular scraping for updated data, probably once a month and/or before running own API server.
- Store scraped data in a database.
- Needs to do the following: 
  - Get all subjects
  - Get all subjects by year and semester
  - Search courses by subject (e.g., COMP SCI), number (e.g., 2103) and course title (e.g., Algorithm Design & Data Structures) 
  - Get course details using course id (e.g, 107592)
  - Get course class list using course id (e.g, 107592)
  - Get class info (like time, size, enrolled, and available seats) by class number (e.g, 11472)

### Why would this feature be useful?

Scrapes course data needed for other projects.

### Additional Notes

**NOTE: https://www.adelaide.edu.au/webteam/dfsearch/2024 was used to obtain some of these links**

**get subjects by year**
e.g. https://courseplanner-api.adelaide.edu.au/api/course-planner-query/v1/?target=/system/SUBJECTS_BY_YEAR/queryx&virtual=Y&year_from=2024&year_to=2024

then

**from each of the "SUBJECT" fields, use this to do course search**
e.g. https://courseplanner-api.adelaide.edu.au/api/course-planner-query/v1/?target=/system/COURSE_SEARCH/queryx&virtual=Y&year=2024&subject=COMP%20SCI&pagenbr=1&pagesize=500

then

**from "COURSE_ID" field, use this to do course detail search**
e.g. https://courseplanner-api.adelaide.edu.au/api/course-planner-query/v1/?target=/system/COURSE_DTL/queryx&virtual=Y&year=2024&courseid=107592&term=4410

then

**use "COURSE_ID" field, use this to get course class list**
e.g. https://courseplanner-api.adelaide.edu.au/api/course-planner-query/v1/?target=/system/COURSE_CLASS_LIST/queryx&virtual=Y&crseid=107592&offer=1&term=4410&session=1

---

**You can search by terms** - use this mainly for determining term codes
e.g. https://courseplanner-api.adelaide.edu.au/api/course-planner-query/v1/?target=/system/TERMS/queryx&virtual=Y&year_from=2024&year_to=2024

---

image

### PR DESCRIPTION
**Pull request by hughmancoder**
### Description
Improved project structure and added scraper functionality to improve readability of code, testability and maintainability.
I've seperated the concerns of the api logic in api/ directory and server/ logic

### Changes Made
- refactor: Refactored and structured project to be more maintainable  (with main 23/09/24)
- feat: Added Test suite to test current classes and establish good testing habits to avoid breaking changes
- feat: Tested example data (https://github.com/compsci-adl/courses-api/issues/2)
- chore: Added .git ignore template 
- docs: Improved README
- chore: Added makefile'

## Related Issue

Fixes #2

### Additional Notes
*Add any other notes or screenshots that might be helpful for the reviewers.*
-tests passing with all example fields listed on issue #2 ticket.
- I've added TODO's for the methods that need testing and have commented out some of the additional api endpoints I have added. I am going on holiday for the next week so perhaps someone else could finish off the testing

### CODE REVIEW

**Comment by phoenixpereira** [comment]
Probably would need to wait until #4 and #5 are merged to do some more refactoring since those PR's would change multiple things.

**Comment by jsun969** [comment]
Please edit your PR title to describe what you did in this PR. TY!

**Comment by hughmancoder** [comment]
> Please edit your PR title to describe what you did in this PR. TY!

- Done: descriptions are in changes made section

**Comment by jsun969** [comment]
It will be better if you follow https://github.com/compsci-adl/.github/blob/main/CONTRIBUTING.md#commit-messages to name your PR title.

**Comment by phoenixpereira** [comment]
@hughmancoder Can you update your branch with main?

**Comment by hughmancoder** [comment]
> @hughmancoder #4 and #5 have now been merged, so can you update your branch with main and make changes as needed?

Done. I've tested all the server routes and can confirm we get a response. Documentation updated too and unit tests passing

**Review by phoenixpereira — CHANGES_REQUESTED** [review]
Some changes needed for formatting and for the scraping to work.

File: src/main.py
Comment (rayokamoto) [inline src/main.py:1]: Why does this exist if it's an empty file?

File: src/db/server.py
Comment (rayokamoto) [inline src/db/server.py:106]: Why are these commented out and not outright removed? Or if there is a chance they'll be used, just leave it uncommented. Generally not a fan of having dead code like this with no explanation.

**Review by rayokamoto — CHANGES_REQUESTED** [review]
In my opinion, the way the files have been split up into `api` and `db` folders do not make sense. This entire project **is** the API. We also don't have that many files so keeping it all under `src` makes sense. If we get way more files then it may make more sense to start splitting it up.

In addition, please format all of your files using Black.

Looks like Phoenix has already given some feedback so I'll just leave it at this for now.

File: README.md
Comment (rayokamoto) [reply README.md:12]: Yes

File: src/db/server.py
Comment (phoenixpereira) [reply src/db/server.py:1]: We can do this in a seperate PR.

**Comment by phoenixpereira** [comment]
@hughmancoder We made some changes to the project structure already to get the scraper working and make the server work with the api, so if you can make a separate PR just for adding testing that would be great.

**Pull request closed without merging**
